### ISSUE TITLE
Intermittent issues establishing a message connection to peers

### ISSUE THREAD

**Issue by jpdillingham** [jpdillingham/Soulseek.NET#234]
The mechanism for establishing connections is currently creating issues establishing a working connection.

Presently when a connection is attempted the code attempts to connect both directly (by establishing a socket directly to the remote user's IP) and indirectly (by sending a `ConnectToPeer` request through the server, causing the peer to connect to our IP).  This works fine when port forwarding is not enabled, preventing indirect connections entirely.

The issue is that under the right, timing-related circumstances when port forwarding is enabled, both direct and indirect connections will succeed.  Depending on the order in which the remote client determines that the connections were established, we may end up with a "bad" connection.

Looking at the logs in the official client, it discards an existing connection when a new connection is established; I have a feeling it is discarding the connection that we believe was established last, which causes the remote client to ignore messages we send.

This isn't a problem when receiving messages; we retain both connections and can receive on both.  This is a problem sending messages.

More research is needed at this point, and I'll continue to talk to myself via comments in this issue to help 🦆

**Comment by jpdillingham** [comment]
Couple of notes:

* I don't have any evidence that this affects transfer connections.  These are single-use per transfer, so there's no cache and no re-use.  That being said, we do have to send an offset upon connection and it is possible that we wind up sending it to a connection that has been discarded, so this is possible _in theory_
* There's a few places we do this "shotgun" style connection negotiation; `PeerConnectionManager` when establishing both unsolicited message and transfer connections, `DistributedConnectionManager` when connecting to parent candidates, and _kinda_ `SoulseekClient.DownloadAsync()` where we wait for both direct and indirect transfer connections to be established.  The latter may not be a concern.

**Comment by jpdillingham** [comment]
There were multiple issues with the design of the `PeerConnectionManager` class.  I hadn't looked in a while, but I confirmed that connection throttling wasn't working, cached connections were being left in a disconnected state, and a bunch of little concurrency issues were just making the entire thing behave in a very non-deterministic way.

I removed connection throttling via `MessageSemaphore`.  This frankly just doesn't work well, elongates searches, causes viable connections to time out while waiting, and when we are capable of accepting incoming connections, will never work to limit them.  Searching for popular terms can cause heavy load; the solution is for users to make a habit out of searching for shorter terms.  I don't like deferring to user behavior to solve issues but there's just no way to do this that doesn't solve worse issues elsewhere.

I removed the semaphore that accompanied cached connections in favor of an implementation over `ConcurrentDictionary` using `Lazy<T>` for value factories.  This has reduced the complexity of the concurrent parts of this class to almost nothing while also working better.
